File: harmony/pushy/src/main/cpp/PushyTurboModule.cpp

#include "PushyTurboModule.h"

using namespace rnoh;
using namespace facebook;

static jsi::Value _hostFunction_PushyTurboModule_getConstants(
    jsi::Runtime &rt,
    react::TurboModule & turboModule,
    const jsi::Value* args,
    size_t count)
    {
        return jsi::Value(static_cast<ArkTSTurboModule &> (turboModule).call(rt,"getConstants", args, count));
    }

static jsi::Value _hostFunction_PushyTurboModule_setLocalHashInfo(
    jsi::Runtime &rt,
    react::TurboModule & turboModule,
    const jsi::Value* args,
    size_t count)
    {
        return jsi::Value(static_cast<ArkTSTurboModule &> (turboModule).call(rt,"setLocalHashInfo", args, count));
    }
    
static jsi::Value _hostFunction_PushyTurboModule_getLocalHashInfo(
    jsi::Runtime &rt,
    react::TurboModule & turboModule,
    const jsi::Value* args,
    size_t count)
    {
        return jsi::Value(static_cast<ArkTSTurboModule &> (turboModule).call(rt,"getLocalHashInfo", args, count));
    }

static jsi::Value _hostFunction_PushyTurboModule_setUuid(
    jsi::Runtime &rt,
    react::TurboModule & turboModule,
    const jsi::Value* args,
    size_t count)
    {
        return jsi::Value(static_cast<ArkTSTurboModule &> (turboModule).call(rt,"setUuid", args, count));
    }

static jsi::Value _hostFunction_PushyTurboModule_reloadUpdate(
    jsi::Runtime &rt,
    react::TurboModule & turboModule,
    const jsi::Value* args,
    size_t count)
    {
        return jsi::Value(static_cast<ArkTSTurboModule &> (turboModule).call(rt,"reloadUpdate", args, count));
    }

static jsi::Value _hostFunction_PushyTurboModule_setNeedUpdate(
    jsi::Runtime &rt,
    react::TurboModule & turboModule,
    const jsi::Value* args,
    size_t count)
    {
        return jsi::Value(static_cast<ArkTSTurboModule &> (turboModule).call(rt,"setNeedUpdate", args, count));
    }

static jsi::Value _hostFunction_PushyTurboModule_markSuccess(
    jsi::Runtime &rt,
    react::TurboModule & turboModule,
    const jsi::Value* args,
    size_t count)
    {
        return jsi::Value(static_cast<ArkTSTurboModule &> (turboModule).call(rt,"markSuccess", args, count));
    }

static jsi::Value _hostFunction_PushyTurboModule_downloadPatchFromPpk(
    jsi::Runtime &rt,
    react::TurboModule & turboModule,
    const jsi::Value* args,
    size_t count)
    {
        return jsi::Value(static_cast<ArkTSTurboModule &> (turboModule).call(rt,"downloadPatchFromPpk", args, count));
    }

static jsi::Value _hostFunction_PushyTurboModule_downloadPatchFromPackage(
    jsi::Runtime &rt,
    react::TurboModule & turboModule,
    const jsi::Value* args,
    size_t count)
    {
        return jsi::Value(static_cast<ArkTSTurboModule &> (turboModule).call(rt,"downloadPatchFromPackage", args, count));
    }

static jsi::Value _hostFunction_PushyTurboModule_downloadFullUpdate(
    jsi::Runtime &rt,
    react::TurboModule & turboModule,
    const jsi::Value* args,
    size_t count)
    {
        return jsi::Value(static_cast<ArkTSTurboModule &> (turboModule).call(rt,"downloadFullUpdate", args, count));
    }

static jsi::Value _hostFunction_PushyTurboModule_downloadAndInstallApk(
    jsi::Runtime &rt,
    react::TurboModule & turboModule,
    const jsi::Value* args,
    size_t count)
    {
        return jsi::Value(static_cast<ArkTSTurboModule &> (turboModule).call(rt,"downloadAndInstallApk", args, count));
    }



static jsi::Value _hostFunction_PushyTurboModule_addListener(
    jsi::Runtime &rt,
    react::TurboModule & turboModule,
    const jsi::Value* args,
    size_t count)
    {
        return jsi::Value(static_cast<ArkTSTurboModule &> (turboModule).call(rt,"addListener", args, count));
    }

static jsi::Value _hostFunction_PushyTurboModule_removeListeners(
    jsi::Runtime &rt,
    react::TurboModule & turboModule,
    const jsi::Value* args,
    size_t count)
    {
        return jsi::Value(static_cast<ArkTSTurboModule &> (turboModule).call(rt,"removeListeners", args, count));
    }

PushyTurboModule::PushyTurboModule(const ArkTSTurboModule::Context ctx, const std::string name)
    : ArkTSTurboModule(ctx,name)
{
    methodMap_["getConstants"]= MethodMetadata{1, _hostFunction_PushyTurboModule_getConstants};
    methodMap_["setLocalHashInfo"]= MethodMetadata{2, _hostFunction_PushyTurboModule_setLocalHashInfo};
    methodMap_["getLocalHashInfo"]= MethodMetadata{3, _hostFunction_PushyTurboModule_getLocalHashInfo};
    methodMap_["setUuid"]= MethodMetadata{1, _hostFunction_PushyTurboModule_setUuid};
    methodMap_["reloadUpdate"]= MethodMetadata{0, _hostFunction_PushyTurboModule_reloadUpdate};
    methodMap_["setNeedUpdate"]= MethodMetadata{0, _hostFunction_PushyTurboModule_setNeedUpdate};
    methodMap_["markSuccess"]= MethodMetadata{0, _hostFunction_PushyTurboModule_markSuccess};
    methodMap_["downloadPatchFromPpk"]= MethodMetadata{0, _hostFunction_PushyTurboModule_downloadPatchFromPpk};
    methodMap_["downloadPatchFromPackage"]= MethodMetadata{0, _hostFunction_PushyTurboModule_downloadPatchFromPackage};
    methodMap_["downloadFullUpdate"]= MethodMetadata{0, _hostFunction_PushyTurboModule_downloadFullUpdate};
    methodMap_["downloadAndInstallApk"]= MethodMetadata{0, _hostFunction_PushyTurboModule_downloadAndInstallApk};
    methodMap_["addListener"]= MethodMetadata{1, _hostFunction_PushyTurboModule_addListener};
    methodMap_["removeListeners"]= MethodMetadata{1, _hostFunction_PushyTurboModule_removeListeners};
}

File: harmony/src/main/ets/PushyTurboModule.ts
import { TurboModule, TurboModuleContext } from '@rnoh/react-native-openharmony/ts';
import common from '@ohos.app.ability.common';
import dataPreferences from '@ohos.data.preferences';
import { bundleManager } from '@kit.AbilityKit';
import abilityAccessCtrl, { Permissions } from '@ohos.abilityAccessCtrl';
import { BusinessError } from '@ohos.base';
import logger from './Logger';
import { UpdateModuleImpl } from './UpdateModuleImpl';
import { UpdateContext } from './UpdateContext';
import { EventHub } from './EventHub';

const TAG = "PushyTurboModule"

export class PushyTurboModule extends TurboModule {
  mUiCtx: common.UIAbilityContext
  context: UpdateContext

  constructor(protected ctx: TurboModuleContext) {
    super(ctx);
    logger.debug(TAG, ",PushyTurboModule constructor");
    this.mUiCtx = ctx.uiAbilityContext
    this.context = new UpdateContext(this.mUiCtx)
    EventHub.getInstance().setRNInstance(ctx.rnInstance)
  }


getConstants(): Object {
  logger.debug(TAG, ",call getConstants");
  const context = this.mUiCtx;
  const preferencesManager = dataPreferences.getPreferencesSync(context,{ name: 'update' });
  const isFirstTime = preferencesManager.getSync("isFirstTime", false) as boolean;
  const rolledBackVersion = preferencesManager.getSync("rolledBackVersion", "") as string;
  const uuid = preferencesManager.getSync("uuid", "") as string;
  const currentVersion = preferencesManager.getSync("currentVersion", "") as string;
  const buildTime = preferencesManager.getSync("buildTime", "") as string;
  const isUsingBundleUrl = this.context.getIsUsingBundleUrl();
  let bundleFlags = bundleManager.BundleFlag.GET_BUNDLE_INFO_WITH_REQUESTED_PERMISSION;
  let packageVersion = '';
  try {
    const bundleInfo = bundleManager.getBundleInfoForSelfSync(bundleFlags);
    packageVersion = bundleInfo?.versionName || "Unknown"
  } catch (error) {
    console.error("Failed to get bundle info:", error);
  }

  if (isFirstTime) {
    preferencesManager.deleteSync("isFirstTime");
  }

  if (rolledBackVersion) {
    preferencesManager.deleteSync("rolledBackVersion");
  }

  return {
    downloadRootDir: `${context.filesDir}/_update`,
    packageVersion,
    currentVersion,
    buildTime,
    isUsingBundleUrl,
    isFirstTime,
    rolledBackVersion,
    uuid,
  }
}


  async setLocalHashInfo(hash: string, info: string): Promise<boolean>  {
    logger.debug(TAG, ",call setLocalHashInfo");
    return UpdateModuleImpl.setLocalHashInfo(this.context,hash,info);
  }

  async getLocalHashInfo(hash: string): Promise<string> {
    return UpdateModuleImpl.getLocalHashInfo(this.context,hash);
  }

  async setUuid(uuid: string): Promise<boolean>  {
    logger.debug(TAG, `,call setUuid`);
    return UpdateModuleImpl.setUuid(this.context,uuid);
  }

  async reloadUpdate(options: { hash: string }): Promise<void> {
    logger.debug(TAG, `,call reloadUpdate`);
    return UpdateModuleImpl.reloadUpdate(this.context, this.mUiCtx, options);
  }

  async setNeedUpdate(options: { hash: string }): Promise<boolean> {
    logger.debug(TAG, `,call setNeedUpdate`);
    return UpdateModuleImpl.setNeedUpdate(this.context, options);
  }

  async markSuccess(): Promise<boolean> {
    logger.debug(TAG, `,call markSuccess`);
    return UpdateModuleImpl.markSuccess(this.context);
  }

  async downloadPatchFromPpk(options: { updateUrl: string; hash: string; originHash: string }): Promise<void> {
    logger.debug(TAG, `,call downloadPatchFromPpk`);
    return UpdateModuleImpl.downloadPatchFromPpk(this.context, options);
  }

  async downloadPatchFromPackage(options: { updateUrl: string; hash: string }): Promise<void>  {
    logger.debug(TAG, `,call downloadPatchFromPackage`);
    return UpdateModuleImpl.downloadPatchFromPackage(this.context, options);
  }

  async downloadFullUpdate(options: { updateUrl: string; hash: string }): Promise<void> {
    logger.debug(TAG, `,call downloadFullUpdate`);
    return UpdateModuleImpl.downloadFullUpdate(this.context, options);
  }

  async downloadAndInstallApk(options: { url: string; target: string; hash: string }): Promise<void> {
    logger.debug(TAG, `,call downloadAndInstallApk`);
    return UpdateModuleImpl.downloadAndInstallApk(this.mUiCtx, options);
  }

  addListener(eventName: string): void {
    logger.debug(TAG, `,call addListener`);
  }

  removeListeners(count: number): void {
    logger.debug(TAG, `,call removeListeners`);
  }
}